### DISCUSSION TITLE
How can I fetch data in next js Component.

### DISCUSSION THREAD

**Discussion by mshshuvooo** [vercel/next.js#21241] [Help]
**getStaticProps()** does not work in Component label rendering. It only works on pages. I have a **Menu** component in my next js app. I want to render data from the server and display the data in the Menu component. If I use getStaticProps() on every page, it doesn't look good. I can't re-use the code for menu rendering. Now how Can I solve the problem? 
Here is my code, I want to re-use this, don't want to copy it in every page. Thanks. 

`<pre>

export async function getStaticProps() {

  const client = new ApolloClient({
    uri: 'http://portfolio.local/graphql',
    cache: new InMemoryCache()
  });

  const { data } = await client.query({
    query: gql`
    query MyQuery {
      menuItems {
        edges {
          node {
            id
            label
          }
        }
      }
    }
    `
  });
  return {
    props: {
      menus: data.menuItems.edges
    }
  }
}
</pre>`

**Comment by lalit2005** [comment]
You can't have a component use getStaticProps() as per docs.
https://nextjs.org/docs/basic-features/data-fetching#only-allowed-in-a-page
You can refer [this discussion](https://github.com/vercel/next.js/discussions/16684)

**Reply by mshshuvooo** [reply]
How can I use getStaticProps in _app.js? Can you please give me an example?

**Reply by toyamarinyon** [reply]
It seems that `getStaticProps` in `_app.js` is not currently supported.

It is discussed in this thread: [https://github.com/vercel/next.js/discussions/10949](https://github.com/vercel/next.js/discussions/10949)

And it seems to be addressed in this PR: [https://github.com/vercel/next.js/issues/8193](https://github.com/vercel/next.js/issues/8193)

However, no specific date has been announced.

**Reply by mshshuvooo** [reply]
@toyamarinyon Hi. Can you please tell me now what can I do for The Menu Component? I am trying to get rid of this problem for the last two days.

**Reply by toyamarinyon** [reply]
@m-allanson Hi.

This is my own sample.
Demo: https://nextjs-dry-menu-sample.vercel.app
Code: https://github.com/toyamarinyon/nextjs-dry-menu-sample
Capture:

https://user-images.githubusercontent.com/535254/104928150-a5dc6780-59e5-11eb-92b5-794ca7af0c09.mov

It may not be exactly what you want, but I think it's one of the more reasonable implementaions using current next.js(10.0.5).
I made this in 30 minutes, so it's messy.

This sample implements the menu with the useEffect hook.(It does not use getStaticProps)

Mocks the communication with the server and waits 2000ms to indicate that loading is complete.(loading... -> loaded!)

I think it works well, but I'd like to improve the fact that every time the page changes, it needs to communicate with the server and waits for 2000ms to load.

So, I made a cached version, because I think caching works well in most cases.

Demo: https://nextjs-dry-menu-sample.vercel.app/with-cache
Capture:

https://user-images.githubusercontent.com/535254/104929787-af66cf00-59e7-11eb-80fd-3baade579e4d.mov

When you visit the first page, it takes 2000ms for the side effects to complete. Then it stores the result in the cache
After the second visit, the side effect will take 500ms to complete because of the cache.  (In fact, I think it's often shorter than that.)

**Reply by toyamarinyon** [reply]
If you want to use `getStaticProps `, then I think https://github.com/vercel/next.js/discussions/21241#discussioncomment-290032 is the current implementation.

As an alternative solution, we introduced a way to read the data on the client side.

**Reply by mshshuvooo** [reply]
@toyamarinyon
Thanks for your example. But your menu is static. My menu is dynamic.  What can I do for this?

**Reply by toyamarinyon** [reply]
Hou about something like this?

Demo: https://nextjs-dry-menu-sample.vercel.app/with-cache
Code: https://github.com/toyamarinyon/nextjs-dry-menu-sample/commit/1532d3666c94023dc2429ad7a6088e12d78fb68e

**Reply by mshshuvooo** [reply]
@toyamarinyon Nice. But If I want to fetch menu data from API, what should I do then?

**Reply by toyamarinyon** [reply]
This is just a sample to introduce the direction of the approach. Therefore, we would like you to think about the specific implementation by yourself.

I've also created a sample that implements the data retrieved from the API up to the point of menuing, so please refer to it.

Demo: https://nextjs-dry-menu-sample.vercel.app/with-cache/contact
Code: https://github.com/toyamarinyon/nextjs-dry-menu-sample/commit/e08a0b4887d1e34ac012cf72c88543521089e637

**Comment by stigkj** [comment]
You extract the code in getStaticProps into a method you call for each page and let the component take props as input.

**Reply by mshshuvooo** [reply]
@stigkj Hi. Thanks a lot for your comment. 
I am very new in Next.js. Can you please give an example. I will be very grateful.

**Comment by megetron** [comment]
just extract the data in the page component.

it will force you to repeat this data requests in each page so this is not the perfect solution i guess.

```
const Home = ({ items }) => {
  return (
   <Menu articles={items}/>
  )
}

export async function getStaticProps() {

  const [items] = await Promise.all([
    fetchAPI('/items?_limit=-1')
  ]);
  return {
    props: { items }
  };
}

export default Home;
```


you can also try to share with context in the _app page if you don't want to repeat. strapi has a nextjs blog which they show an example so you can follow their code in [here](https://github.com/strapi/strapi-starter-next-blog/blob/master/frontend/pages/_app.js)

**Reply by mshshuvooo** [reply]
@megetron  Thank you. So if I want to fetch menu data by using getStaticProps(), then I have to repeat getStaticProps on every page, am I right? 

And the example of Strapi they use getInitialProps() Instead of getStaticProps(). But Next.js recommends to use getStaticProps(), but It does not work on _app.js. On the other hand, getInitialProps() works fine on _app.js.

**Reply by stigkj** [reply]
@mshshuvooo It might be that the pages won't be static when adding a `getInitialProps` to `_app.js`. I don't have time to check this out now, but just try it out and you can see it in the output from `next build`.
